Mise à jour FullCalendar v4 -> v5

francois-ferrandis opened the pull request:
[Review app](https://demo-rdv-solidarites-pr5187.osc-secnum-fr1.scalingo.io/) - `martine@demo.rdv-solidarites.fr` - `Rdvservicepublictest1!`

# Pourquoi mettre à jour ?

Plusieurs raisons :
- Nous prévoyons de customiser notre agenda de façon poussée, notamment pour l'affichage de plages d'ouverture. La v5 de FullCalendar : 
  - permet une customisation plus poussée
  - simplifie son markup HTML
  - fournit des outils dédiés à la customisation (la v7 ira encore plus loin, voir https://github.com/fullcalendar/fullcalendar/issues/7907)
- la future version 7 de FullCalendar apportera notamment des amélioration d'accessibilité ([voir leur roadmap ici](https://fullcalendar.io/roadmap))
- la version actuelle est la v4 n'est plus supportée. La v5 que l'on atteint ici est tagguée "legacy" [sur npm](https://www.npmjs.com/package/fullcalendar) et n'a reçu aucune MAJ depuis mai 2023. La v6, tagguée "latest", est la seule à recevoir des bugfixes : le passage de la v5 à la v6 est très facile pour nous et se fera dans une PR séparée.


# Comment j'ai mis à jour

## La théorie

Le guide de migration officiel est très instructif : 
https://fullcalendar.io/docs/v5/upgrading-from-v4

On y décrit les changements suivants : 

- **Renommage d'options** : Comme on le voit dans le diff de cette PR, certaines options sont renommées, par exemple `defaultView` devient `initialView`.
- **Changement de structure HTML / CSS** : plus d'infos sur ces adaptations dans la section "Le CSS" ci-dessous
- **Le JS de FullCalendar importe lui-même son CSS** (j'ignorais que c'était possible), donc plus besoin d'imports explicites dans `application_agent.scss`
- De nouvelles pratiques pour customiser un calendrier (injection de classes CSS custom et de contenu, ainsi que l'introduction de [variables CSS](https://github.com/fullcalendar/fullcalendar/blob/v5/packages/common/src/styles/vars.css)) : j'utilise désormais l'option `eventClassNames` pour injecter nos classes custom. J'ai laissé le reste des customisation là où elles étaient.

 ## La pratique

# Le CSS

C'est le plus gros du travail sur cette PR, car il a fallu non seulement trouver quelles customisations CSS étaient cassées par la MAJ, mais aussi déterminer quelles parties de `_calendar.scss` étaient inopérantes déjà avant de mettre à jour. 

Idéalement j'aurais dû faire du ménage dans une PR séparée avant cette PR de mise à jour, mais j'ai avancé à tâtons et je pense que ça coûterait trop cher de séparer en commits propres après coup. Je demande donc ici un peu de confiance. J'ai ajouté en commentaire des explications pour certains changements.

Il y a des petites différences cosmétiques entre avant et après, j'ai visé la simplicité du code puisqu'on aura le loisir de customiser lourdement (avec beaucoup de commentaires) dans les semaines qui viennent.

### Renommage de nos classes custom

J'ai fait en sorte que les classes que nous rajoutons commencent par `.rdv-`, afin de ne pas les confondre avec les classes natives de FullCalendar qui sont préfixées par `.fc-`.

# Captures écran


Avant | Après
:- | -:
![image](https://github.com/user-attachments/assets/6e193e3d-b4ee-4711-b7d1-5531c10a10d5) | ![image](https://github.com/user-attachments/assets/15cc0d3a-e8da-4f25-aae8-5e8c4358a0b9)
![image](https://github.com/user-attachments/assets/48aee58a-bcea-473d-97cd-fdcb8c9663b5) | ![image](https://github.com/user-attachments/assets/72a1f179-13d8-4d95-ab1b-b12434b127df)
![image](https://github.com/user-attachments/assets/1d7a72ea-572f-4472-a100-54b8f676186b) | ![image](https://github.com/user-attachments/assets/924e339e-25f6-4cdf-8ff1-98f76b5ce5ae)
![image](https://github.com/user-attachments/assets/602ae740-9c0a-47f3-bfa6-89430ae0f972) | ![image](https://github.com/user-attachments/assets/fb20ae0a-d7b4-42ce-93c1-ad0f37177d84)

# Fonctionnalités à conserver

J'ai vérifié que les fonctionnalités ci-dessous sont bien toujours fonctionnelles après cette migration.

## Indication d'usager en salle d'attente 

![image](https://github.com/user-attachments/assets/9eb037bb-c0da-43de-b4ed-bd1c7ff55b20)

## Indication de RDV annulé

![image](https://github.com/user-attachments/assets/736fe2d9-29a2-4672-adbc-5f9e381bd62f)

## RDV présent sur l'agenda d'un collègue mais inaccessible pour moi

![image](https://github.com/user-attachments/assets/52aa9c85-725e-4877-bf36-2c1dc8e82e45)

# Améliorations repérées durant cette MAJ

-  ~~Les RDVs honorés sont affichés de la même façon que les RDVs annulés (barrés et transparents), ce que j'ai trouvé surprenant car je m'attendais à ce qu'ils soient affichés avec un :heavy_check_mark: par exemple plutôt que barrés.~~ Je me suis trompé, en réalité les RDVs annulés sont barrés et transparents et les honorés sont simplement transparents, tout va bien !

francois-ferrandis committed:
MAJ des packages de 4.4.2 vers 5.11.5

francois-ferrandis committed:
Retire les imports CSS désormais plus nécessaires

francois-ferrandis committed:
Renomme les propriétés

https://fullcalendar.io/docs/v5/upgrading-from-v4

francois-ferrandis committed:
Je ne trouve aucune trace de ".none-border"

francois-ferrandis committed:
On gagne un peu en espace et en lisibilité je trouve

francois-ferrandis committed:
WIP

francois-ferrandis committed:
Franchement plein de trucs

francois-ferrandis committed:
Ça commence à ressembler à avant

francois-ferrandis committed:
On s'approche

francois-ferrandis committed:
Renomme .selected en .rdv-shake (approche "utility class")

francois-ferrandis committed:
Ajout de commentaire

francois-ferrandis committed:
CSS lint

francois-ferrandis committed:
Revert d'un changement cosmétique non sollicité

francois-ferrandis committed:
Adapte le JS côté RdvPlan

francois-ferrandis committed:
Ne pas afficher l'heure de fin

francois-ferrandis committed:
Ajustements d'affichage pour coller à la version actuelle

francois-ferrandis committed:
Ajout de commentaire

francois-ferrandis committed:
Préférer une class utilitaire plutôt que des status dynamiques

francois-ferrandis committed:
Renomme les classes custom et utilise le hook

francois-ferrandis committed:
Centrer verticalement le contenu des longs événements

francois-ferrandis commented on app/javascript/components/calendar/utils.js:
```diff
@@ -71,9 +76,7 @@ function eventRenderer(selectedEventId) {
     }
 
-    if (info.event.rendering == 'background') {
-      $el.append("<div class=\"fc-title\" style=\"color: white; padding: 2px 4px; font-size: 12px; font-weight: bold;\">" + info.event.title + "</div>");
```
Cette ligne n'est plus nécessaire car désormais le titre des background events s'affiche par défaut.

francois-ferrandis commented on app/javascript/stylesheets/components/_calendar.scss:
```diff
@@ -1,135 +1,78 @@
-.fc-view {
-  margin-top: 30px;
-}
```
Je vire un peu de margin pour sauver les agents qui ont un petit écran, je trouve que le spacing reste correct.

francois-ferrandis commented on app/javascript/stylesheets/components/_calendar.scss:
```diff
@@ -1,135 +1,78 @@
+$rdvs-fc-header-bg-color: $gray-200;
 
-.none-border {
-  .modal-footer {
```
Je n'ai trouvé cette classe `.modal-footer` nulle part.

francois-ferrandis commented on app/javascript/stylesheets/components/_calendar.scss:
```diff
@@ -1,135 +1,78 @@
   height: auto;
-}
-
-.fc-button:hover {
-  color: $black;
-}
```
Inutile maintenant qu'on a des variables CSS

francois-ferrandis commented on app/javascript/stylesheets/components/_calendar.scss:
```diff
@@ -1,135 +1,78 @@
-  color: $black;
-}
-
-.fc-text-arrow {
-  font-family: inherit;
-  font-size: 1rem;
-}
-
-.fc-state-hover {
-  background: $gray-200;
-}
-
-.fc-state-highlight {
-  background: $gray-300;
-}
-
-.fc-state-down,
-.fc-state-active,
-.fc-state-disabled {
-  background-color: $primary;
-  color: $white;
-  text-shadow: none;
-}
```
Je n'ai pas trouvé l'utilité de ces définitions, elles datent de 2019, je préfère jeter ! :broom:

francois-ferrandis commented on app/javascript/stylesheets/components/_calendar.scss:
```diff
@@ -1,135 +1,78 @@
-.fc-event-excused,
-.fc-event-seen,
-.fc-event-revoked {
+.rdv-fc-event-barre {
```
Ces trois classes étaient ajoutées dynamiquement donc difficile à retrouver.  J'ai remplacé par une classe utilitaire `.rdv-fc-event-barre`.

francois-ferrandis commented on app/javascript/stylesheets/components/_calendar.scss:
```diff
@@ -144,101 +87,54 @@
   }
 }
 
-@include media-breakpoint-down(sm) {
-  .fc-toolbar {
-    .fc-left,
-    .fc-right,
-    .fc-center {
-      float: none;
-      display: block;
-      clear: both;
-      margin: 10px 0;
-    }
-  }
-  .fc {
-    .fc-toolbar {
-      > * {
-        > * {
-          float: none;
-        }
-      }
-    }
-  }
-  .fc-today-button {
-    display: none;
-  }
-}
-
```
Je n'ai pas trouvé quel était l'effet de cette horreur.

AntoineGirard commented on app/javascript/stylesheets/components/_calendar.scss:
```diff
@@ -1,135 +1,78 @@
-.fc-view {
-  margin-top: 30px;
-}
+$rdvs-fc-header-bg-color: $gray-200;
```
Merci beaucoup pour l’ajout des commentaires dans ce fichier (et le travail de recherche et de nettoyage) c’est beaucoup plus clair)

AntoineGirard reviewed (approved):
Merci beaucoup François ! 

Testé en review app et ça me semble good

> Les RDVs honorés sont affichés de la même façon que les RDVs annulés (barrés et transparents), ce que j'ai trouvé surprenant car je m'attendais à ce qu'ils soient affichés avec un ✔️ par exemple plutôt que barrés.

Effectivement, c'est étrange cette partie. À la limite, si c’était transparent, mais pas barré, ça me semblerait déjà + logique. Mais je pense qu’on peut se garder ça pour + tard.

francois-ferrandis committed:
Définit une config par défaut

francois-ferrandis commented:
> Effectivement, c'est étrange cette partie. À la limite, si c’était transparent, mais pas barré, ça me semblerait déjà + logique. Mais je pense qu’on peut se garder ça pour + tard.

En fait je m'étais trompé, les honorés sont en transparent mais pas barrés, my bad ! :facepalm:

adipasquale reviewed (approved):
🙇 👏 🚀 🎸 

Énorme merci pour ce gros boulot minutieux @francois-ferrandis 
je suis fan des nouveaux préfixes de classe CSS et des commentaires 🙇 

J’ai parcouru le code rapidement je n’ai rien à redire

Sur le JS qui embarque le CSS est-ce qu’on est sûrs que c’est compatible avec tous les navigateurs / contextes ? peut-être qu’on peut inspecter un peu les requêtes faites pour voir si ça en rajoute ? et je suis curieux de comprendre le mécanisme précisément. 

j’ai joué au jeu des 7 différences, j’en ai trouvé qques unes sur la vue planning surtout

<img width="1922" alt="image" src="https://github.com/user-attachments/assets/6c944e0f-5338-414e-b22f-672c9a1af835" />

rien de bien grave

francois-ferrandis committed:
Corrige la couleur des headers

francois-ferrandis committed:
Réduit la taille du texte uniquement des events de grid

francois-ferrandis committed:
Cacher l'heure de fin sur les événements courts uniquement

francois-ferrandis committed:
Conserver l'affichage en gras des plages

francois-ferrandis committed:
Rétablit notre CSS custom sur les pages RdvPlan

francois-ferrandis committed:
Corrige l'affichage du curseur

francois-ferrandis committed:
Merge branch 'production' into frf/full-calendar-v5

francois-ferrandis committed:
CSS Lint

francois-ferrandis committed:
Corrige la hauteur de l'affichage en jour du RdvPlan

francois-ferrandis committed:
Affiche également un "pointer" sur les plages d'ouverture

francois-ferrandis committed:
Corrige la police dans le tooltip

francois-ferrandis committed:
Ce code crée des bugs lorsqu'on switch entre la vue semaine et jour

Je préfère que l'on affiche toujours l'heure de fin, quitte à ajouter une
option pour les agents "ne pas afficher l'heure de fin", qu'ils peuvent
toggle à volonté selon leurs besoins.

francois-ferrandis committed:
viewDidMount n'est pas appelé lorsqu'on switch entre timeGridWeek et timeGridOneDay

francois-ferrandis committed:
Pourquoi ne pas stocker la liste comme dernière vue choisie ?

francois-ferrandis committed:
Les classes FC ne semblent plus s'appliquer en dehors, let's go DSFR

francois-ferrandis commented on app/views/admin/agent_agendas/show.html.slim:
```diff
@@ -28,14 +28,12 @@ ruby:
-       = f.hidden_field :display_cancelled_rdv, value: !current_agent.display_cancelled_rdv
-       / The fc-button fc-button-primary classes give the button the aspect of calendar buttons
-       = f.submit Agent.human_attribute_value(:display_cancelled_rdv, !current_agent.display_cancelled_rdv), class: "fc-button fc-button-primary", style: "text-transform: none;"
+      = f.hidden_field :display_cancelled_rdv, value: !current_agent.display_cancelled_rdv
+      = f.submit Agent.human_attribute_value(:display_cancelled_rdv, !current_agent.display_cancelled_rdv), class: "fr-btn fr-btn--tertiary"
```
L'application des classes `fc-button fc-button-primary` ne fonctionne plus, probablement à cause du chargement dynamique du CSS. Je suis passé au DSFR, tant pis pour la cohérence, de toute façon ces boutons sont à mettre dans une section "Paramètres du calendrier", avec de nouvelles options comme "Afficher l'heure de fin sur les événements courts".

adipasquale replied:
👍 c’est nickel merci pour le commentaire

![image](https://github.com/user-attachments/assets/561d4917-c734-4882-af8b-2d4b42c37e66)

adipasquale commented on app/javascript/stylesheets/rdv_plan.scss:
```diff
@@ -13,3 +15,9 @@
 }
+
+// Besoin de forcer cette déclaration pour utiliser Marianne dans
+// le tooltip, sinon c'est Liberation Sans qui est utilisée.
```
bien vu 👍

adipasquale reviewed (approved):
👍 merci pour tout ce boulot minutieux @francois-ferrandis 

J’ai refait un tour en local je n’ai rien à redire.

Ce qui change le plus c’est la vue mois. Ça me paraît plus lisible dans la nouvelle version mais potentiellement certaines personnes vont se plaindre, on pourra réagir et réappliquer un style plus proche de l’ancien si on le souhaite ça sera pas trop dur ni urgent à mon avis. 

<img width="1074" alt="image" src="https://github.com/user-attachments/assets/813dd1f8-b721-4acf-879a-b832b954c674" />

francois-ferrandis committed:
Merge branch 'production' into frf/full-calendar-v5
